utils/firestore.js

```diff
@@ -2502,78 +2502,136 @@ const setPackageReceiptUSPS = async (data) => {
+                });
+            });
+        });
+
+        let batch = db.batch();
+        const receivedTimestamp = boxUpdateData[fieldMapping.shipmentReceivedTimestamp];
+
+        for (const key in collectionIdHolder) {
+            const specimenDoc = specimenDocs.find(specimenDoc => specimenDoc.data[fieldMapping.collectionId] === key);
+            const specimenDocRef = specimenDoc.ref;
+            const specimenData = specimenDoc.data;
 
-    for (const key in collectionIdHolder) {
-        try {
             let updateObject = {};
-            const secondSnapshot = await db.collection("biospecimen").where('820476880', '==', collectionIdHolder[key]).get(); // find related biospecimen using collection id
-            const docId = secondSnapshot.docs[0].id;
-            const specimenData = secondSnapshot.docs[0].data();
 
-            updateObject['926457119'] = dateTimeStamp;
+            // only update the specimen level received timestamp if it is not already set. Important so stray tubes don't update the specimen level received timestamp.
+            if (!specimenData[fieldMapping.shipmentReceivedTimestamp]) {
+                updateObject[fieldMapping.shipmentReceivedTimestamp] = receivedTimestamp;
+            }
 
-            for (const element of collectionIdKeys[key]['234868461']) {
-                //grab tube ids & map them to appropriate concept ids. If it's a misc tube (0050-0054), find tube's location in specimen to get the concept id.
+            // Map tube ids to concept ids. If misc tube (0050-0054), find tube's location in specimen to get the concept id.
+            for (const element of collectionIdHolder[key]) {
                 const tubeId = element.split(' ')[1];
-                
-                let conceptTube = collectionIdConversion[tubeId]; 
+
+                let tubeConceptId = collectionIdConversion[tubeId]; 
                 if (miscTubeIdSet.has(tubeId)) {
-                    conceptTube = Object.keys(specimenData).find(tubeKey => tubeConceptIds.includes(tubeKey) && specimenData[tubeKey][fieldMapping.objectId] === element);
+                    tubeConceptId = Object.keys(specimenData).find(tubeKey => tubeConceptIds.includes(tubeKey) && specimenData[tubeKey][fieldMapping.objectId] === element);
                 }
 
-                if (conceptTube) {
-                    const conceptIdTubes = `${conceptTube}.926457119`
-                    updateObject[conceptIdTubes] = dateTimeStamp;
+                if (!tubeConceptId) {
+                    console.error('tube concept ID not found for', element);
+                } else {
+                    const conceptIdTubes = `${tubeConceptId}.926457119`;
+                    updateObject[conceptIdTubes] = receivedTimestamp;
                 }
             }
-            await db.collection("biospecimen").doc(docId).update(updateObject);
+            console.log(collectionIdHolder[key], 'updateObject', updateObject);
```

Comment:
Was this console.log left in on purpose?

Reply:
Nope, removed. Thanks!